XRC: use wxBitmapBundle in wxBookCtrl-related controls.

kosh543 opened the pull request.

kosh543 committed:
XRC: use wxBitmapBundle in wxBookCtrl-related controls.

kosh543 committed:
fixup! XRC: use wxBitmapBundle in wxBookCtrl-related controls.

kosh543 committed:
Fix usage of wxWithImages::Images in wxTreebook

kosh543 committed:
XRC: use wxBitmapBundle in wxTreebook

vadz commented on include/wx/treebook.h:
```diff
@@ -132,6 +132,7 @@ class WXDLLIMPEXP_CORE wxTreebook : public wxNavigationEnabled<wxBookCtrlBase>
     virtual int HitTest(const wxPoint& pt, long *flags = NULL) const wxOVERRIDE;
+    virtual void SetImages(const Images& images) wxOVERRIDE;
     virtual void SetImageList(wxImageList *imageList) wxOVERRIDE;
     virtual void AssignImageList(wxImageList *imageList);
```
We should override `OnImagesChanged()` instead and remove all the other ones:
```suggestion
    virtual void OnImagesChanged() wxOVERRIDE;
```

vadz commented on include/wx/withimages.h:
```diff
@@ -65,7 +65,7 @@ class WXDLLIMPEXP_CORE wxWithImages
     // Override OnImagesChanged() in the derived class to update the actually
     // shown images.
-    void SetImages(const Images& images)
+    virtual void SetImages(const Images& images)
```
This contradicts the comment just above and shouldn't be done.
```suggestion
    void SetImages(const Images& images)
```

vadz commented on include/wx/xrc/xh_choicbk.h:
```diff
@@ -27,6 +27,8 @@ class WXDLLIMPEXP_XRC wxChoicebookXmlHandler : public wxXmlResourceHandler
     bool m_isInside;
     wxChoicebook *m_choicebook;
+    wxVector<wxBitmapBundle> m_bookImages;
+    wxVector<size_t> m_bookImagesIdx;
```
It would be nice to explain somewhere what does this vector hold. Of course, the not very nice thing is that it's used in several places and we don't want to repeat the same description in all of them, so ideal would be to add some `wxBookCtrlXmlResourceHandlerBase` class and put this field and the above one in into it and write this comment just once there.

vadz commented on src/generic/treebkg.cpp:
```diff
@@ -564,6 +564,12 @@ wxWindow *wxTreebook::TryGetNonNullPage(size_t n)
     return page;
 }
 
+void wxTreebook::SetImages(const Images& images)
```
We need to do here the same thing as in `wxListbook::OnImagesChanged()`, except we don't need to care about `wxLC_ICON` style, as we always use the icons here if they're specified.

vadz commented on src/xrc/xh_choicbk.cpp:
```diff
@@ -123,9 +116,25 @@ wxObject *wxChoicebookXmlHandler::DoCreateResource()
         m_choicebook = nb;
         bool old_ins = m_isInside;
         m_isInside = true;
+        wxVector<wxBitmapBundle> old_images = m_bookImages;
+        m_bookImages.clear();
```
This would be slightly more efficient (also below and at the of the function):
```suggestion
        wxVector<wxBitmapBundle> old_images;
        swap(m_bookImages, old_images);
```

kosh543 committed:
fixup! Fix usage of wxWithImages::Images in wxTreebook

vadz commented on src/xrc/xh_choicbk.cpp:
```diff
@@ -123,9 +116,25 @@ wxObject *wxChoicebookXmlHandler::DoCreateResource()
+        m_bookImagesIdx.clear();
         CreateChildren(m_choicebook, true/*only this handler*/);
+
+        if ( !m_bookImages.empty() )
+        {
+            m_choicebook->SetImages(m_bookImages);
+            for ( size_t i = 0; i < m_bookImagesIdx.size(); ++i )
+            {
+                m_choicebook->SetPageImage( m_bookImagesIdx[i], i );
+            }
+        }
```
This could be refactored into `wxBookCtrlXmlResourceHandlerBase::DoSetPageImages()` to avoid duplicating it in many places.

vadz committed:
fixup! Fix usage of wxWithImages::Images in wxTreebook

vadz committed:
Factor out common code from most wxBookCtrl-related XRC handlers

Add new wxBookCtrlXmlHandlerBase and inherit the XRC handlers of all the
wxBookCtrlBase-derived classes except for wxTreebookXmlHandler, which
will require special handling, from it to avoid duplicating the same
code in all of them.

This commit is best viewed with --color-moved git option.

vadz committed:
Move wxBookCtrlXmlHandlerBase::PageWithAttrs into the header

This struct is finally going to have to be declared in the header as it
will be needed by wxTreebookXmlHandler when it's modified to derive from
wxBookCtrlXmlHandlerBase too.

Also add GetImageId() function to the struct while moving it.

vadz committed:
Add virtual wxBookCtrlXmlHandlerBase::DoAddPage()

This will allow customizing the way pages are added to the book control
and will also be used in wxTreebookXmlHandler soon.

vadz committed:
Inherit wxTreebookXmlHandler from wxBookCtrlXmlHandlerBase too

Use the same logic for creating the pages of this book control as for
all the other ones and keep only wxTreebook-specific "depth" and
"expanded" nodes handling in this class.

vadz commented:
I've added my changes implementing the previous remarks. I have a couple of other clean ups in mind, but they shouldn't change much.

@kosh543 Could you please review and, ideally, test this and tell me if you see anything wrong with these changes? TIA!

vadz committed:
Don't declare wxPropertySheetDialogXmlHandler if !wxUSE_BOOKCTRL

This class is only implemented when wxUSE_BOOKCTRL==1 as its code uses
it, so it shouldn't be declared when wxUSE_BOOKCTRL==0 neither.

This fixes wxiOS build problem after the latest changes.

vadz commented:
Rebased and merged in the commit mentioned above, thanks.
